README.md
# HIPPO Gym 
##### Human Input Parsing Platform for Openai Gym

Written by [Nick Nissen](https://nicknissen.com) and Yuan Wang
Supervised by [Matt Taylor](https://drmatttaylor.net) and Neda Navi
For the Intelligent Robot Learning Laboratory [(IRLL)](https://irll.ca) at the University of Alberta [(UofA)](https://ualberta.ca)
Supported by the Alberta Machine Intelligence Institute [(AMII)](https://amii.ca)

For questions or support contact us at [hippogym.irll@gmail.com](mailto:hippogym.irll@gmail.com)

The HIPPO Gym Project contains 3 repositories:

1. The main framework: [HIPPO_Gym](https://github.com/IRLL/HIPPO_Gym)

2. The AWS code and instructions: [HIPPO_Gym_AWS](https://github.com/IRLL/HIPPO_Gym_AWS)

3. The React Front End: [HIPPO_Gym_React_FrontEnd](https://github.com/IRLL/HIPPO_Gym_FrontEnd_React)

For members of the IRLL, CMPUT656, or anyone whose organization has already setup the AWS infrastructure, the only repo required is #1.

# Real time and Muti-modal Software Framework for Reinforcement Learning

What behind our framework
===============
This project is related to the researh where a human can help teach a software agent to learn faster.For example, you have someone play a game of Mario, and use this demonstration to help the agent to learn. Or, you have someone watch an agent playing Mario and interrupt to say "you should have jumped here". Or, you have someone watch an agent playing Mario and interrupt to say "you just messed up here." In the past, we've focused on doing things locally - building an interface for these kind of interactions isn't so bad when you just have someone sit down in front of a laptop. But, ideally, we'd be able to run these kinds of experiments over the web.

What our framework can do
===============
* Interact with Turkers recruited via Amazon’s Mechanical Turk, as well as human subjects recruited from other venues
* Support multiple forms of human guidance, including explicit reward, action advice, demonstrations, and shaping reward function construction.
* Handle bi-directional requests, where guidance can be initiated by either the human teacher or the student agent. 
* Allow for real-time interaction with a live agent, as well as pre-recorded agent behaviors.
* Allow a person to go forward and backward in time to provide human guidance at different points in previous trajectories
* Be compatible with at least 3 tasks implemented in OpenAI gym.

How it works
===============

## Demo Video
[HIPPO Gym Demo Video](https://youtu.be/KeMzhGU9LWE) (Updated on 2021-04-28)

In this demo video, we provided no other parameters in the original query. There are two pages before the game page and two pages after. The first page is a consent form that requires the user to agree the agreement before they move to the next step. In the second step, there is a qualification test to see if the user meet all the requirements for this project. The next page is the game page which is also the main page that used to play the game and train the agent. After the game ends, there is a survey page ask about the whole experience of this framework. The last page is a "thank you" page to thank the research participants.

## API
### 1.Get the first pre-game page through ``https://api.irll.net/next/`` GET
```
params : {
    #Id assigned to different projects
    "projectId" : ,
    #Generated by uuidv4
    "userId" : 
}
```
Response status code and body format:

**200 OK**
```
{   
    # html content body
    "page" : ,
    # css content to apply on the page
    "css" :  ,
}
```
**400 Bad Request**
```
{
    "Project ID Not Found"
}
```
### 2.Submit a form and get the next pre-game page through ``https://api.irll.net/next`` POST
Request format
```
params : {
        #Id assigned to different projects
        "projectId" : ,
        #Generated by uuidv4
        "userId" : 
    }
form content example
{
    "race" : "Asian",
    "country" : "Canada",
    "language" : "English"
}
```
#### Response status code and body format:
**200 OK**
```
{
    # html content body
    "page" : ,
    # available css styles that applies on the page 
    "css" :  ,
}
```

## URL Query parameters
* projectId(**required**) : projectId is assigned by each researcher, and it is uniquely identifiers a project
* css(**optional**) : An url that points to a public css file on the internet. This field is used if you want to apply your own styles to this framework.
* server(**optional**) : An url that points to a back end server. This field is used if you don't want to use the default back end server.
* debug(**optional**) : This mode is used to display the incoming and outgoing messages on the page, it is not enabled by default. In most of the time, the researchers will not need to use this mode. 
* redirect(**optional**) : This field is used to set the redirect link after the game ends.
* userId(**optional**) : Unique uuid used to uniquely identifiers an user.

Getting started
===============

The following instructions will get you a copy of this project and you can run the project on your local machine.

### Prerequisites

You need to install the following software:

* Node

* npm

### Clone

* Clone this repo to your local machine
*  `git clone https://github.com/IRLL/HIPPO_Gym_FrontEnd_React.git`


### Structure
    
[main.js](./src/main.js)
```javascript
fetchFormData()
```
Function used to send **GET** requests to our api endpoint and receive the html contents of pre-game and post-game pages.
```javascript
gameEndHandler()
```
Function used to set the game state to be end and request the content of next page.
```javascript
handleSubmit(event)
```
Function used to handle the form data after the user click the "Submit" button. This function will send POST requests to the back end and set the response as the content of the next page.

[Routes.js](./src/Routes.js)
A router used to manage all the routes of our framework. Currently, we only have one route which is the root one("/"). In the future, we may introduce more routes to our framework.

[forum.js](./src/components/Forum/forum.js)
Forum component, used to render the html contents that fetched from the back end and includes a "Submit" button to submit the form's data.

[game.js](./src/components/Game/game.js)
Game componet that include all the elements that needed to play the game on the web. In this component, we are using websocket to communicate with our back end server, which makes our framework real time. There are two main parts in this component, the first is the game window. We render every game frames that received from the web socket in this window. The second part is the game control panel, we have 4 directions button and buttons to start, stop and stop the game, etc.
```javascript
handleOk()
handleCancel()
```
Functions used to handle the visibility of the confirmation model.By default, the model does not appear, it is visible to the user when the game ends.
```javascript
sendMessage(data)
```
Function used to send data to the websocket server in JSON format.
```javascript
handleCommand(status)
```
Function used to send essential game control commands to the websocket server. The commands we have are : start,pause,stop and reset, etc.
```javascript
handleFPS(speed)
```
Function used to control the FPS(frame per second) of the game. The default value is 30 and the increment or the decrement when each time the user press the button is 5. The allowed range of the FPS is between 5 - 90 for now.

[gameWindow.js](./src/components/GameWindow/gameWindow.js)
Game window component, a component used to render every frames from the server. Before the game starts, there is a loading component inside the window. After the game is ready to start, the first frame of the game will be rendered.

[control.js](./src/components/Control/control.js)
ControlPanel component includes all the elements that neeeded to control the game. For example, we have direction control buttons and "start", "stop","reset", etc.
We have introduced the selective UI to our ControlPanel to meet different requirements for each project. Researcher can select the control buttons they want for their project.

**Control components we have right now:**
- Game control
    - Left
    - LeftUp
    - LeftDown
    - Right
    - RightUp
    - RightDown
    - Up
    - Down
    - Fire
- Train Control
    - Start
    - Pause
    - Stop
    - Reset
    - Train Online
    - Train Offline
    - Good
    - Bad
    - Fps Up
    - Fps Down
    - Fps Set
    
[budgetBar.js](./src/components/BudgetBar/budgetBar.js)
Budget Bar component, used to render the usedInputBudget and inputBudget. This component is visible if there are limits on the number of times that the user can give the feedback by clicking the "good" and "bad" buttons.

[displayBar.js](./src/components/DisplayBar/displayBar.js)
Display Bar component, used to render extra inforamtion that the researcher want to display on the game page. For example, the researcher can choose to display the score of the last game, the number of total episodes and the remaining episodes.

[error.js](./src/components/Error/error.js)
Error component, this component is only visible to the user if there are errors in the query parameters.

[MessageViewer.js](./src/components/Message/MessageViewer.js)
Message viewer is used when debug mode is set to true in the query parameters. By enabling the debug mode, all incoming and outgoing messages will be showed on the page.
 
[header.js](./src/components/Layout/customHeader.js)
Header component that will be rendered on every pages of our framework. We only have a irll logo and a step bar to indicate the current step for now.

[footer.js](./src/components/Layout/customFooter.js)
Footer component that will be rendered on every pages of our framework. We have logos for ualberta and amii.

#### Setup

> Install the packages required for frontend 

```shell
$ npm install 
```

### Run

> Run frontend in the browser

```shell
$ npm start
```

### Build

> Build the code for production
```
$ npm run build
```

### Dependencies

* [react](https://www.npmjs.com/package/react) 
> The main Javascript library we are using for building user interfaces.
* [react-icons](https://www.npmjs.com/package/react-icons)
> npm package that allows us to include only the popular icons that our project is using.
* [react-device-detect](https://www.npmjs.com/package/react-device-detect)
> npm package that used to detect the type of device of user for better user experience
* [react-html-parser](https://www.npmjs.com/package/react-html-parser)
> npm package that used to convert HTML strings into React components. Avoids the use of dangerouslySetInnerHTML and converts standard HTML elements, attributes and inline styles into their React equivalents.
* [react-router](https://www.npmjs.com/package/react-router)
> npm package that used to provide routing in React.
* [antd](https://www.npmjs.com/package/antd)
> Ant-design is an enterprise-class UI design language and React UI library.Most of our framework's UI design based on this library.
* [websocket](https://www.npmjs.com/package/websocket)
> npm package that used to build web-socket connection with our backend server. We are relying on web-socket to receive the game frames and communicating between the front end and the back end.
* [axios](https://www.npmjs.com/package/axios)
> npm package that used to send GET and POST request to our api endpoints
* [uuid](https://www.npmjs.com/package/uuid)
> npm package used to generate uuid(universally unique identifier). We use uuidv4 in our project as the userId to identify each user uniquelly.
* [react-helmet](https://www.npmjs.com/package/react-helmet)
> npm package  that is a reusable React component will manage all of your changes to the document head.


changelog.md
# Changelog

## [Unreleased]
### Added
- Support for fingerprint project
- Controls for brightness, contrast, saturation and hue of images
- Images can be zoomed and panned now using the [vxzoom](https://vx-demo.vercel.app/docs/zoom) library
- Option to undo and redo changes made to images
- A minimap has been added to the fingerprint window
- Clicking anywhere within the window provides the coordinates of the click
- Fingerprint window supports markers to mark minutaie
- Markers can be labeled, rotated, resized and colored
- Game screen can now toggle between horizontal and vertical views
- Changelog has been added and is formatted based on [this site](https://keepachangelog.com/en/1.0.0/)

### Changed
- Changed the the gamewindow and control panel to have thinner borders
- style changes for header, footer and main page
- frame rate can now be changed through manual input in the input textbox

### Fixed
- UI for debug mode is fixed. Debug mode will not allow to toggle orientation

src/components/game.js
import React from 'react';
import 'antd/dist/antd.css';
import './game.css';
import {message, Modal, Row, Col } from 'antd';
import { w3cwebsocket } from "websocket";
import {browserName,osName,browserVersion,osVersion} from 'react-device-detect';
import getKeyInput from '../utils/getKeyInput';
import {WS_URL, USER_ID, PROJECT_ID, SERVER, DEBUG} from '../utils/constants';
import ControlPanel from './control';
import BudgetBar from './budgetBar';
import MessageViewer from './MessageViewer';
import GameWindow from './gameWindow';

const pendingTime = 30;

class Game extends React.Component{
    
    state = {
        frameCount : 0,
        frameId : 0,
        frameRate : 30,
        frameSrc : "",
        isLoading : !SERVER ? true : false,
        isEnd : false,
        isConnection : false,
        isVisible : false,
        UIlist : [],
        progress : 0,
        allData : null,
        inputBudget : 0,
        usedInputBudget : 0,
        receiveData : null,
        isPause : false,
        inMessage : [],
        outMessage : []
    }

    componentDidMount() {

        if(DEBUG){
            this.setInMessage = setInterval(() => {
                if(this.state.receiveData && !this.state.isPause){
                    this.setState(prevState => ({
                        inMessage : [prevState.receiveData,...prevState.inMessage]
                    }))
                }
            },1000)
        }

        //Running a check every 1/100 second(10 millisecond)
        //If allData is not null then send the message
        //otherwise just wait until next checking
        this.sendData = setInterval(() => {
            if(this.state.allData && this.state.isConnection){
                this.websocket.send(JSON.stringify(this.state.allData));
                //record every message send to the server
                if(DEBUG){
                    this.setState(prevState => ({
                        outMessage : [prevState.allData,...prevState.outMessage],
                    }))
                }
                this.setState(({allData : null}));
            }
        }, 10);

        //To update the progress of loading game content
        //Since we always need to wait 30 seconds before the game
        //content get loaded, we update the progress (100/30) per second
        this.updateProgress = setInterval(() => 
            this.setState(prevState => ({
                progress : prevState.progress+(100/pendingTime)
            }))
        ,1000)
        //To ensure the websocket server is ready to connect
        //we try to connect the webscoket server periodically
        //for every 30 seconds until the connection has been established
        this.timer = setTimeout(() => {
            //connect the websocket server
            this.websocket = new w3cwebsocket(WS_URL);
            this.websocket.onopen = () => {
                //Once the websocket connection has been established
                //we remove all the unnecessary timer
                clearTimeout(this.timer);
                clearInterval(this.updateProgress);
                console.log('WebSocket Client Connected');
                this.setState(({
                    isLoading : false,
                    isConnection : true
                }))
                this.sendMessage({
                    userId : USER_ID,
                    projectId : PROJECT_ID
                })
            };

            //listen to the data from the websocket server
            this.websocket.onmessage = (message) => {
                //"done" means the game has ended
                if(message.data === "done"){
                    this.setState(({
                        isEnd : true,
                        isVisible : true
                    }))
                //parse the data from the websocket server
                }else{
                    let parsedData = JSON.parse(message.data);
                    //Check if budget bar should be loaded
                    if(parsedData.inputBudget && parsedData.usedInputBudget){
                        this.setState(({
                            inputBudget : parsedData.inputBudget,
                            usedInputBudget : parsedData.usedInputBudget
                        }))
                    }
                    //Check if filed UI in response
                    if(parsedData.UI){
                        this.setState(({
                            UIlist : parsedData.UI
                        }))
                    }
                    //Check if field frame in response
                    if(parsedData.frame && parsedData.frameId){
                        let frame = parsedData.frame;
                        let frameId = parsedData.frameId;
                        this.setState(prevState => ({
                            frameSrc : "data:image/jpeg;base64, " + frame,
                            frameCount : prevState.frameCount + 1,
                            frameId : frameId
                        }));
                    }
                    //record every message received from the server
                    if(DEBUG){
                        this.setState(({
                            receiveData : parsedData
                        }))
                    }
                }
            };

            //listen to the websocket closing status
            this.websocket.onclose = () => {
                console.log('WebSocket Client Closed');
                this.setState(({
                    isConnection : false
                }))
            }
        }, SERVER ? 0 : pendingTime * 1000);

        //listen to the user's keyboard inputs
        document.addEventListener('keydown', (event) => {
            //Used to prevent arrow keys and space key from scrolling the page
            let dataToSend = getKeyInput(event.code);
            if(dataToSend.actionType !== null){
                event.preventDefault();
            }
            
            if(this.state.UIlist.includes(dataToSend.action)){
                this.sendMessage(dataToSend);
            }
        })
    }

    componentWillUnmount() {
        clearInterval(this.sendData);
        if(this.setInMessage) clearInterval(this.setInMessage);
    }

    //change the confirmation modal to be invisible
    //navigate to the post-game page
    handleOk = e => {
        this.setState({
          isVisible : false
        });
        this.props.action()
      };
    
    //change the confirmation modal to be invisible
    //stay on the game page
    handleCancel = e => {
        this.setState({
            isVisible : false
        });
    };

    //send data to websocket server in JSON format
    sendMessage = (data) => {
        if(this.state.isConnection){
            const allData = {
                ...data,
                frameCount : this.state.frameCount,
                frameId : this.state.frameId
            }
            this.setState(({
                allData : allData
            }))
        }
    }

    //send game control commands to the websocket server
    handleCommand = (status) => {

        const {isLoading, inputBudget, usedInputBudget} = this.state;

        if(isLoading){
            message.error("Please wait the connection to be established first!")
            return;
        }
        if(status === "start"){
            this.sendMessage({
                command : status,
                system : osName,
                systemVersion : osVersion,
                browser : browserName,
                browserVersion : browserVersion,
            })
        }else{
            if(status === 'pause') this.setState(prevState => ({isPause : !prevState.isPause}));
            if(["good","bad"].includes(status) && inputBudget > 0){
                if(usedInputBudget <= inputBudget){
                    this.setState(prevState => ({
                        usedInputBudget : prevState.usedInputBudget + 1
                    }))
                }else{
                    message.error("You have consumed all the reward budget!",3);
                    return;
                }
            }
            this.sendMessage({
                command : status
            })
        }   
    }

    //chnage the FPS of the game
    handleFPS = (speed) => {
        if((speed === "faster" && this.state.frameRate + 5 > 90) || (speed === "slower" && this.state.frameRate - 5 < 1)){
            message.error("Invalid FPS, the FPS can only between 1 - 90!")
        }else{
            this.setState(prevState => ({
                frameRate : speed === "faster" ? prevState.frameRate + 5 : prevState.frameRate - 5
            }))
            this.sendMessage({
                changeFrameRate : speed
            })
        }
    }

    render() {
        const {inMessage, outMessage, isLoading, frameSrc, frameRate, isEnd, UIlist, progress, isVisible, inputBudget, usedInputBudget} = this.state;

        return (
            <div>
                <Row>
                    <Col flex={1}><MessageViewer title="Message In" data={inMessage} visible={DEBUG} /></Col>

                    <Col flex={2}><GameWindow isLoading={isLoading} frameSrc={frameSrc} progress={progress} /></Col>

                    <Col flex={1}><MessageViewer title="Message Out" data={outMessage} visible={DEBUG} /></Col>
                </Row>
                
                <BudgetBar visible={inputBudget>0} isLoading={isLoading} usedInputBudget={usedInputBudget} inputBudget={inputBudget} /> 
                
                <ControlPanel 
                    isEnd={isEnd} 
                    isLoading={isLoading}
                    frameRate={frameRate} 
                    UIlist={UIlist} 
                    handleOk={this.handleOk} 
                    handleFPS={this.handleFPS}
                    handleCommand={this.handleCommand} 
                    sendMessage={this.sendMessage}
                />

                <Modal
                    title="Game end message"
                    visible={isVisible}
                    onOk={this.handleOk}
                    onCancel={this.handleCancel}
                    >
                    <p className="modal">The game has ended</p>
                    <p className="modal">Press <b>"Cancel"</b> to stay on this page</p>
                    <p className="modal">Press <b>"OK"</b> to move to next step</p>
                </Modal>
            </div>
        )
    }
}

export default Game;